barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp

```diff
@@ -14,114 +15,215 @@ template class DSLBigInts<MegaCircuitBuilder>;
 
 template <typename Builder>
 void build_constraints(Builder& builder,
-                       AcirFormat const& constraint_system,
+                       AcirFormat& constraint_system,
                        bool has_valid_witness_assignments,
-                       bool honk_recursion)
+                       bool honk_recursion,
+                       bool collect_gates_per_opcode)
 {
+    if (collect_gates_per_opcode) {
+        constraint_system.gates_per_opcode.resize(constraint_system.num_acir_opcodes, 0);
+    }
+    size_t prev_gate_count = 0;
+
+    auto compute_gate_diff = [&]() -> size_t {
+        if (!collect_gates_per_opcode) {
+            return 0;
+        }
+        size_t new_gate_count = builder.get_num_gates();
+        size_t diff = new_gate_count - prev_gate_count;
+        prev_gate_count = new_gate_count;
+        return diff;
+    };
+
+    auto track_gate_diff = [&](std::vector<size_t>& gates_per_opcode, size_t opcode_index) -> void {
+        if (collect_gates_per_opcode) {
+            gates_per_opcode[opcode_index] = compute_gate_diff();
+        }
+    };
+
     // Add arithmetic gates
-    for (const auto& constraint : constraint_system.poly_triple_constraints) {
+    for (size_t i = 0; i < constraint_system.poly_triple_constraints.size(); ++i) {
+        const auto& constraint = constraint_system.poly_triple_constraints[i];
         builder.create_poly_gate(constraint);
+        track_gate_diff(constraint_system.gates_per_opcode,
+                        constraint_system.original_opcode_indices.poly_triple_constraints[i]);
     }
-    for (const auto& constraint : constraint_system.quad_constraints) {
+
+    for (size_t i = 0; i < constraint_system.quad_constraints.size(); ++i) {
+        const auto& constraint = constraint_system.quad_constraints[i];
         builder.create_big_mul_gate(constraint);
+        track_gate_diff(constraint_system.gates_per_opcode,
+                        constraint_system.original_opcode_indices.quad_constraints[i]);
     }
 
     // Add logic constraint
-    for (const auto& constraint : constraint_system.logic_constraints) {
+    for (size_t i = 0; i < constraint_system.logic_constraints.size(); ++i) {
+        const auto& constraint = constraint_system.logic_constraints[i];
         create_logic_gate(
             builder, constraint.a, constraint.b, constraint.result, constraint.num_bits, constraint.is_xor_gate);
+        track_gate_diff(constraint_system.gates_per_opcode,
+                        constraint_system.original_opcode_indices.logic_constraints[i]);
     }
 
     // Add range constraint
-    for (const auto& constraint : constraint_system.range_constraints) {
+    for (size_t i = 0; i < constraint_system.range_constraints.size(); ++i) {
+        const auto& constraint = constraint_system.range_constraints[i];
```

Comment:
using `.at(i)` is in general preferred over `[i]` unless speed is absolutely critical. In debug (or assert) modes, the former does some basic range checking, while the latter segfaults